python/example_code/cognito/cognito_idp_actions.py

```diff
@@ -0,0 +1,405 @@
+                else:
+                    raise RuntimeError(
+                        "The user pool requires MFA setup, but the user pool is not "
+                        "configured for TOTP MFA. This example requires TOTP MFA.")
+        except ClientError as err:
+            logger.error(
+                "Couldn't start sign in for %s. Here's why: %s: %s", user_name,
+                err.response['Error']['Code'], err.response['Error']['Message'])
+            raise
+        else:
+            response.pop('ResponseMetadata', None)
+            return response
+    # snippet-end:[python.example_code.cognito-idp.AdminInitiateAuth]
+
+    # snippet-start:[python.example_code.cognito-idp.AssociateSoftwareToken]
+    def get_mfa_secret(self, session):
+        """
+        Gets a token that can be used to associate an MFA application with the user.
+
+        :param session: Session information returned from a previous call to initiate
+                        authentication.
+        :return: An MFA token that can be used to set up an MFA application.
+        """
+        try:
+            response = self.cognito_idp_client.associate_software_token(Session=session)
+        except ClientError as err:
+            logger.error(
+                "Couldn't get MFA secret. Here's why: %s: %s",
+                err.response['Error']['Code'], err.response['Error']['Message'])
+            raise
+        else:
+            response.pop('ResponseMetadata', None)
+            return response
+    # snippet-end:[python.example_code.cognito-idp.AssociateSoftwareToken]
+
+    # snippet-start:[python.example_code.cognito-idp.VerifySoftwareToken]
+    def verify_mfa(self, session, user_code):
+        """
+        Verify a new MFA application that is associated with a user.
+
+        :param session: Session information returned from a previous call to initiate
+                        authentication.
+        :param user_code: A code generated by the associated MFA application.
+        :return: Status that indicates whether the MFA application is verified.
+        """
+        try:
+            response = self.cognito_idp_client.verify_software_token(
+                Session=session, UserCode=user_code)
+        except ClientError as err:
+            logger.error(
+                "Couldn't verify MFA. Here's why: %s: %s",
+                err.response['Error']['Code'], err.response['Error']['Message'])
+            raise
+        else:
+            response.pop('ResponseMetadata', None)
+            return response
+    # snippet-end:[python.example_code.cognito-idp.VerifySoftwareToken]
+
+    # snippet-start:[python.example_code.cognito-idp.AdminRespondToAuthChallenge]
+    def respond_to_mfa_challenge(self, user_name, session, mfa_code):
+        """
+        Responds to a challenge for an MFA code. This completes the second step of
+        a two-factor sign in. When sign in is successful, it returns an access token
```

Comment:
both mentions of "sign in" --> "sign-in"

Reply:
fixed